pulse2percept/models/granley2021.py

```diff
@@ -0,0 +1,559 @@
+"""`BiphasicAxonMapModel`"""
+from attr import Attribute
+import numpy as np
+from copy import deepcopy
+import sys
+
+from numpy.core.fromnumeric import size
+
+from . import AxonMapModel, AxonMapSpatial, TemporalModel, Model
+from ..implants import ProsthesisSystem, ElectrodeArray
+from ..stimuli import BiphasicPulseTrain, Stimulus
+from ..datasets import load_nanduri2012
+from ..percepts import Percept
+from ..utils import FreezeError
+
+from ._granley2021 import fast_biphasic_axon_map
+from .base import NotBuiltError
+
+
+class DefaultBrightModel():
+    """
+    Default model to be used for brightness scaling in BiphasicAxonMapModel
+    Implements Eq 4 from [Granley2021]_ 
+    Fit using data from [Nanduri2012]_ and [Weitz2015]_
+
+    Parameters:
+    ------------
+    do_thresholding : bool, optional
+        Set to true to enable probabilistic phosphene appearance at near-threshold 
+        amplitudes
+    **kwargs : dict, optional 
+        Allows for setting model coefficients, A0-4
+    """
+
+    def __init__(self,  do_thresholding=False, **kwargs):
```

Comment:
If you make this class derive from `BaseModel`, you can inherit all the default value and `setattr` logic. Then you don't need a constructor, you just need to supply a `get_default_params` method that returns a dict `{'a0': 0.27, ...}`. Deriving from `BaseModel` also gives you all the `Frozen` logic, where params can't be added outside init/build

Reply:
In fact, I believe this should help you solve the issue of setting parameter values before build. It would no longer be `DefaultBrightModel`, just `BrightModel` I guess. You'd call it like `BrightModel()` to use the default params, or `BrightModel(a0=1)` etc. to override stuff